mlx/backend/metal/fast_inference_sdpa.cpp

```diff
@@ -0,0 +1,236 @@
+  compute_encoder->dispatchThreadgroups(grid_dims, group_dims);
+
+  {
+    auto kernel_accum = d.get_kernel(kname_reduce.str());
+    compute_encoder->setComputePipelineState(kernel_accum);
+    set_array_buffer(compute_encoder, o_partial, 0);
+    set_array_buffer(compute_encoder, p_lse, 1);
+    set_array_buffer(compute_encoder, p_rowmaxes, 2);
+    compute_encoder->setBytes(&params, sizeof(MLXFastInferenceSDPAParams), 3);
+    set_array_buffer(compute_encoder, out, 4);
+
+    MTL::Size grid_dims_reduce = MTL::Size(heads, 1, batch);
+    MTL::Size group_dims_reduce = MTL::Size(128, 1, 1);
+
+    compute_encoder->dispatchThreadgroups(grid_dims_reduce, group_dims_reduce);
+
+    d.get_command_buffer(s.index)->addCompletedHandler(
+        [&, temporaries](MTL::CommandBuffer*) mutable { temporaries.clear(); });
+    return;
+  }
+}
+} // namespace
+
+void FastInferenceSDPA::eval_gpu(const std::vector<array>& inputs, array& out) {
+  assert(inputs.size() >= 3);
+  if (!is_floating_point(out.dtype())) {
+    throw std::runtime_error(
+        "[FastInferenceSDPA] Does not yet support non-floating point types.");
+  }
+
+  if (inputs.size() == 4) {
+    printf("Received 4 inputs\n");
+    throw std::runtime_error(
+        "[FastInferenceSDPA] Mask expected to be null at FastInferenceSDPA::eval_gpu.");
+  }
+
+  out.set_data(allocator::malloc_or_wait(out.nbytes()));
+  auto& s = stream();
+  auto& d = metal::device(s.device);
+
+  auto& q_pre = inputs[0];
+  auto& k_pre = inputs[1];
+  auto& v_pre = inputs[2];
+  auto& o = out;
+  /////////////////////////////////////////////////////////////////////////////
+  // Init checks and prep
+
+  // Keep a vector with copies to be cleared in the completed buffer to release
+  // the arrays
+  std::vector<array> temporaries;
+  auto check_transpose = [&temporaries, &s](const array& arr) {
+    auto stx = arr.strides()[arr.ndim() - 2];
+    auto sty = arr.strides()[arr.ndim() - 1];
+    if (stx == arr.shape(-1) && sty == 1) {
+      return std::make_tuple(false, stx, arr);
+    } else if (stx == 1 && sty == arr.shape(-2)) {
+      return std::make_tuple(true, sty, arr);
+    } else {
+      array arr_copy(arr.shape(), arr.dtype(), nullptr, {});
+      copy_gpu(arr, arr_copy, CopyType::General, s);
+      temporaries.push_back(arr_copy);
+      size_t stx = arr.shape(-1);
+      return std::make_tuple(false, stx, arr_copy);
+    }
+  };
+
+  auto [q_transposed, q_cols, q] = check_transpose(q_pre);
+  auto [k_transposed, k_cols, k] = check_transpose(k_pre);
+  auto [v_transposed, v_cols, v] = check_transpose(v_pre);
+
+  if (q_transposed) {
+    throw std::runtime_error(
+        "[fast_inference_sdpa] q_transposed NYI for Metal backend.");
+  }
+  if (k_transposed) {
+    throw std::runtime_error(
+        "[fast_inference_sdpa] k_transposed NYI for Metal backend.");
+  }
+  if (v_transposed) {
+    throw std::runtime_error(
+        "[fast_inference_sdpa] v_transposed NYI for Metal backend.");
+  }
```

Comment:
I think (?) you could remove this by removing the case where we do the copy. So basically L158 gets removed and it will copy everything to be row-contiguous.

Reply:
done, yes, should be safe.  I also never ran in to the transposed case in practice.